Windows changed readers

dengert opened the pull request:
Possible solution to #2517  It needs testing, and is being submitted as a PR to get Windows artifacts.

    Windows is using new reader with card
    
    Issue #2517 reports that minidriver may be presented with new handles
    to PCSC, even if card is moved from one reader to another.
    The previous code assumed this would never happen, assuming
    that would be two operations: card is removed from one
    reader then inserted at some time later inserted into another.
    
    pcsc_use_reader now tests if the reader name has changed and will return
    SC_ERROR_READER so minidriver can call reinit_card to will cleanup
    and then start using the new reader.
    
    The assumption is if card is removed, it could be modified on another system
    then inserted on original system with changed data. 
    

<!--
Thank you for your pull request.

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX'
(without quotes) in the commit message.

Mention which card(s) are used during testing. To get the name of your card,
run this command: `opensc-tool -n`
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] Documentation is added or updated
- [ ] New files have a LGPL 2.1 license statement
- [ ] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

jozsefd commented:
I have applied the patch to my working copy, and tried to read the certificates from the card. Without removing the card, an error occurs after cca. 20 calls of check_reader_status(), it reports a HANDLES CHANGES and fails:
```
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] MD_Function:CardGetProperty:6357 called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] 
P:3228 T:10316 pCardData:006BFD68 
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] CardGetProperty('Card Identifier',cbData=16,dwFlags=0) called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] check_reader_status for CardGetProperty
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] MD_Function:check_card_status:396 called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] MD_Function:check_card_status:406 returning with: 0x00000000
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] sizeof(size_t):4 sizeof(ULONG_PTR):4 sizeof(__int3264):4 sizeof pCardData->hSCardCtx:4
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] pCardData->hSCardCtx:0xCD010003 hScard:0xEA020000
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] HANDLES CHANGED from  vs->hSCardCtx:0xCD010003 vs->hScard:0xEA010000
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] ctx.c:895:sc_ctx_use_reader: called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] reader-pcsc.c:2532:pcsc_use_reader: called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] Try Reusing the reader
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] Reader name changed from "REINER SCT cyberJack RFID komfort HUN USB 1" to "REINER SCT cyberJack RFID komfort HUN USB 1"
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] reader-pcsc.c:2600:pcsc_use_reader: returning with: -1100 (Generic reader error)
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] sc_ctx_use_reader returned -1100
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] disassociate_card
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] sc_pkcs15_unbind
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] pkcs15.c:1317:sc_pkcs15_unbind: called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] pkcs15-pin.c:863:sc_pkcs15_pincache_clear: called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] sc_disconnect_card
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] card.c:414:sc_disconnect_card: called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] card-nqApplet.c:186:nqapplet_finish: called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] card-nqApplet.c:191:nqapplet_finish: returning with: 0 (Success)
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] card.c:436:sc_disconnect_card: returning with: 0 (Success)
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] MD_Function:associate_card:7151 called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] ctx.c:895:sc_ctx_use_reader: called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] reader-pcsc.c:2532:pcsc_use_reader: called
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] Try Reusing the reader
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] Reader name changed from "REINER SCT cyberJack RFID komfort HUN USB 1" to "REINER SCT cyberJack RFID komfort HUN USB 1"
P:3228; T:10316 2022-03-04 10:10:45.475 [cardmod] reader-pcsc.c:2600:pcsc_use_reader: returning with: -1100 (Generic reader error)
P:3228; T:10316 2022-03-04 10:10:45.490 [cardmod] sc_ctx_use_reader() failed with -1100
P:3228; T:10316 2022-03-04 10:10:45.490 [cardmod] MD_Function:associate_card:7171 returning with: 0x8010002F
P:3228; T:10316 2022-03-04 10:10:45.490 [cardmod] MD_Function:reinit_card:342 returning with: 0x8010002F
P:3228; T:10316 2022-03-04 10:10:45.490 [cardmod] CardGetProperty was called, but unable to initialize card, r = 2148532271
P:3228; T:10316 2022-03-04 10:10:45.490 [cardmod] MD_Function:reinit_card_for:389 returning with: 0x8010002F
```

I will double check it with the artifact MSI soon.

frankmorgner commented on src/libopensc/reader-pcsc.c:
```diff
@@ -2548,7 +2562,7 @@ int pcsc_use_reader(sc_context_t *ctx, void * pcsc_context_handle, void * pcsc_c
 	if(SCARD_S_SUCCESS == gpriv->SCardGetAttrib(card_handle,
 				SCARD_ATTR_DEVICE_SYSTEM_NAME_A, (LPBYTE)
-				reader_name, &reader_name_size)) {
+				&reader_name, &reader_name_size)) {
```
Adding `&` here is wrong

dengert replied:
It is actually wrong in 2 places, at line 2543, should not have & in both.

dengert commented:
@jozsefd The MSI with fix suggested by Frank https://github.com/OpenSC/OpenSC/pull/2518/commits/1189f9dfedb8fbb23526fe07aae01300b017802a are building now, they are found on this page under "Checks", "continuous-integration/appveyor/pr" and clicking on "Details" 

But I see another bug, `|| strcmp(reader_name, reader->name) == 0) {` should use "!="

Jakuje commented on src/libopensc/reader-pcsc.c:
```diff
@@ -2523,21 +2523,35 @@ int pcsc_use_reader(sc_context_t *ctx, void * pcsc_context_handle, void * pcsc_c
+				SCARD_ATTR_DEVICE_SYSTEM_NAME_A, (LPBYTE)
+				reader_name, &reader_name_size)
+				|| strcmp(reader_name, reader->name) != 0) {
+			sc_log( ctx,"Reader name changed from \"%s\" to \"%s\"", reader->name, reader_name);
```
```suggestion
			sc_log(ctx, "Reader name changed from \"%s\" to \"%s\"", reader->name, reader_name);
```

jozsefd commented:
Sorry for the delay. I tested the current version of this PR, it fixes the problems above, but somehow it still fails if I remove the card from ReinerSCT and plug into the Omnikey:

```
P:18220; T:17028 2022-03-08 16:16:39.146 [cardmod] CardReadFile
P:18220; T:17028 2022-03-08 16:16:39.147 [cardmod] pszDirectoryName = mscp, pszFileName = cmapfile, dwFlags = 0, pcbData=0301EB88, ppbData=0301EB8C
P:18220; T:17028 2022-03-08 16:16:39.147 [cardmod] check_reader_status for CardReadFile
P:18220; T:17028 2022-03-08 16:16:39.147 [cardmod] MD_Function:check_card_status:396 called
P:18220; T:17028 2022-03-08 16:16:39.147 [cardmod] MD_Function:check_card_status:406 returning with: 0x00000000
P:18220; T:17028 2022-03-08 16:16:39.148 [cardmod] sizeof(size_t):4 sizeof(ULONG_PTR):4 sizeof(__int3264):4 sizeof pCardData->hSCardCtx:4
P:18220; T:17028 2022-03-08 16:16:39.148 [cardmod] pCardData->hSCardCtx:0xCD030004 hScard:0xEA020001
P:18220; T:17028 2022-03-08 16:16:39.148 [cardmod] HANDLES CHANGED from  vs->hSCardCtx:0xCD030002 vs->hScard:0xEA040000
P:18220; T:17028 2022-03-08 16:16:39.148 [cardmod] ctx.c:895:sc_ctx_use_reader: called
P:18220; T:17028 2022-03-08 16:16:39.149 [cardmod] reader-pcsc.c:2527:pcsc_use_reader: called
P:18220; T:17028 2022-03-08 16:16:39.149 [cardmod] Try Reusing the reader
P:18220; T:17028 2022-03-08 16:16:39.149 [cardmod] Reader name changed from "REINER SCT cyberJack RFID komfort HUN USB 1" to "OMNIKEY AG Smart Card Reader USB 0"
P:18220; T:17028 2022-03-08 16:16:39.150 [cardmod] reader-pcsc.c:2595:pcsc_use_reader: returning with: -1100 (Generic reader error)
P:18220; T:17028 2022-03-08 16:16:39.150 [cardmod] sc_ctx_use_reader returned -1100
P:18220; T:17028 2022-03-08 16:16:39.150 [cardmod] disassociate_card
P:18220; T:17028 2022-03-08 16:16:39.150 [cardmod] sc_pkcs15_unbind
P:18220; T:17028 2022-03-08 16:16:39.151 [cardmod] pkcs15.c:1317:sc_pkcs15_unbind: called
P:18220; T:17028 2022-03-08 16:16:39.151 [cardmod] pkcs15-pin.c:863:sc_pkcs15_pincache_clear: called
P:18220; T:17028 2022-03-08 16:16:39.151 [cardmod] sc_disconnect_card
P:18220; T:17028 2022-03-08 16:16:39.151 [cardmod] card.c:414:sc_disconnect_card: called
P:18220; T:17028 2022-03-08 16:16:39.152 [cardmod] card-nqApplet.c:186:nqapplet_finish: called
P:18220; T:17028 2022-03-08 16:16:39.152 [cardmod] card-nqApplet.c:191:nqapplet_finish: returning with: 0 (Success)
P:18220; T:17028 2022-03-08 16:16:39.152 [cardmod] card.c:436:sc_disconnect_card: returning with: 0 (Success)
P:18220; T:17028 2022-03-08 16:16:39.152 [cardmod] MD_Function:associate_card:7151 called
P:18220; T:17028 2022-03-08 16:16:39.153 [cardmod] ctx.c:895:sc_ctx_use_reader: called
P:18220; T:17028 2022-03-08 16:16:39.153 [cardmod] reader-pcsc.c:2527:pcsc_use_reader: called
P:18220; T:17028 2022-03-08 16:16:39.153 [cardmod] Try Reusing the reader
P:18220; T:17028 2022-03-08 16:16:39.153 [cardmod] Reader name changed from "REINER SCT cyberJack RFID komfort HUN USB 1" to "OMNIKEY AG Smart Card Reader USB 0"
P:18220; T:17028 2022-03-08 16:16:39.154 [cardmod] reader-pcsc.c:2595:pcsc_use_reader: returning with: -1100 (Generic reader error)
P:18220; T:17028 2022-03-08 16:16:39.154 [cardmod] sc_ctx_use_reader() failed with -1100
P:18220; T:17028 2022-03-08 16:16:39.154 [cardmod] MD_Function:associate_card:7171 returning with: 0x8010002F
P:18220; T:17028 2022-03-08 16:16:39.155 [cardmod] MD_Function:reinit_card:342 returning with: 0x8010002F
P:18220; T:17028 2022-03-08 16:16:39.155 [cardmod] CardReadFile was called, but unable to initialize card, r = 2148532271
P:18220; T:17028 2022-03-08 16:16:39.155 [cardmod] MD_Function:reinit_card_for:389 returning with: 0x8010002F
```

It causes a crash in basecsp.dll. I am going to analyse this scenario and try to find the reason of it.

dengert commented:
OK that debug log shows the first reader was not cleaned up so second call
use_reader tried to reuse it again.

But was it with or without the diff I sent?

Will look at it when my internet comes back up.


On Tue, Mar 8, 2022, 11:14 AM Jozsef Dojcsak ***@***.***>
wrote:

> Sorry for the delay. I tested the current version of this PR, it fixes the
> problems above, but somehow it still fails if I remove the card from
> ReinerSCT and plug into the Omnikey:
>
> P:18220; T:17028 2022-03-08 16:16:39.146 [cardmod] CardReadFile
> P:18220; T:17028 2022-03-08 16:16:39.147 [cardmod] pszDirectoryName = mscp, pszFileName = cmapfile, dwFlags = 0, pcbData=0301EB88, ppbData=0301EB8C
> P:18220; T:17028 2022-03-08 16:16:39.147 [cardmod] check_reader_status for CardReadFile
> P:18220; T:17028 2022-03-08 16:16:39.147 [cardmod] MD_Function:check_card_status:396 called
> P:18220; T:17028 2022-03-08 16:16:39.147 [cardmod] MD_Function:check_card_status:406 returning with: 0x00000000
> P:18220; T:17028 2022-03-08 16:16:39.148 [cardmod] sizeof(size_t):4 sizeof(ULONG_PTR):4 sizeof(__int3264):4 sizeof pCardData->hSCardCtx:4
> P:18220; T:17028 2022-03-08 16:16:39.148 [cardmod] pCardData->hSCardCtx:0xCD030004 hScard:0xEA020001
> P:18220; T:17028 2022-03-08 16:16:39.148 [cardmod] HANDLES CHANGED from  vs->hSCardCtx:0xCD030002 vs->hScard:0xEA040000
> P:18220; T:17028 2022-03-08 16:16:39.148 [cardmod] ctx.c:895:sc_ctx_use_reader: called
> P:18220; T:17028 2022-03-08 16:16:39.149 [cardmod] reader-pcsc.c:2527:pcsc_use_reader: called
> P:18220; T:17028 2022-03-08 16:16:39.149 [cardmod] Try Reusing the reader
> P:18220; T:17028 2022-03-08 16:16:39.149 [cardmod] Reader name changed from "REINER SCT cyberJack RFID komfort HUN USB 1" to "OMNIKEY AG Smart Card Reader USB 0"
> P:18220; T:17028 2022-03-08 16:16:39.150 [cardmod] reader-pcsc.c:2595:pcsc_use_reader: returning with: -1100 (Generic reader error)
> P:18220; T:17028 2022-03-08 16:16:39.150 [cardmod] sc_ctx_use_reader returned -1100
> P:18220; T:17028 2022-03-08 16:16:39.150 [cardmod] disassociate_card
> P:18220; T:17028 2022-03-08 16:16:39.150 [cardmod] sc_pkcs15_unbind
> P:18220; T:17028 2022-03-08 16:16:39.151 [cardmod] pkcs15.c:1317:sc_pkcs15_unbind: called
> P:18220; T:17028 2022-03-08 16:16:39.151 [cardmod] pkcs15-pin.c:863:sc_pkcs15_pincache_clear: called
> P:18220; T:17028 2022-03-08 16:16:39.151 [cardmod] sc_disconnect_card
> P:18220; T:17028 2022-03-08 16:16:39.151 [cardmod] card.c:414:sc_disconnect_card: called
> P:18220; T:17028 2022-03-08 16:16:39.152 [cardmod] card-nqApplet.c:186:nqapplet_finish: called
> P:18220; T:17028 2022-03-08 16:16:39.152 [cardmod] card-nqApplet.c:191:nqapplet_finish: returning with: 0 (Success)
> P:18220; T:17028 2022-03-08 16:16:39.152 [cardmod] card.c:436:sc_disconnect_card: returning with: 0 (Success)
> P:18220; T:17028 2022-03-08 16:16:39.152 [cardmod] MD_Function:associate_card:7151 called
> P:18220; T:17028 2022-03-08 16:16:39.153 [cardmod] ctx.c:895:sc_ctx_use_reader: called
> P:18220; T:17028 2022-03-08 16:16:39.153 [cardmod] reader-pcsc.c:2527:pcsc_use_reader: called
> P:18220; T:17028 2022-03-08 16:16:39.153 [cardmod] Try Reusing the reader
> P:18220; T:17028 2022-03-08 16:16:39.153 [cardmod] Reader name changed from "REINER SCT cyberJack RFID komfort HUN USB 1" to "OMNIKEY AG Smart Card Reader USB 0"
> P:18220; T:17028 2022-03-08 16:16:39.154 [cardmod] reader-pcsc.c:2595:pcsc_use_reader: returning with: -1100 (Generic reader error)
> P:18220; T:17028 2022-03-08 16:16:39.154 [cardmod] sc_ctx_use_reader() failed with -1100
> P:18220; T:17028 2022-03-08 16:16:39.154 [cardmod] MD_Function:associate_card:7171 returning with: 0x8010002F
> P:18220; T:17028 2022-03-08 16:16:39.155 [cardmod] MD_Function:reinit_card:342 returning with: 0x8010002F
> P:18220; T:17028 2022-03-08 16:16:39.155 [cardmod] CardReadFile was called, but unable to initialize card, r = 2148532271
> P:18220; T:17028 2022-03-08 16:16:39.155 [cardmod] MD_Function:reinit_card_for:389 returning with: 0x8010002F
>
> It causes a crash in basecsp.dll. I am going to analyse this scenario and
> try to find the reason of it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenSC/OpenSC/pull/2518#issuecomment-1062011793>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGTIMON3DVAYZA6NR5R6FDU66DJHANCNFSM5P25META>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

dengert commented on src/libopensc/reader-pcsc.c:
```diff
@@ -2523,21 +2523,35 @@ int pcsc_use_reader(sc_context_t *ctx, void * pcsc_context_handle, void * pcsc_c
+			goto out;
+		}
+
+		priv->pcsc_card =*(SCARDHANDLE *)pcsc_card_handle;
```
fixed

dengert commented on src/libopensc/reader-pcsc.c:
```diff
@@ -2523,21 +2523,35 @@ int pcsc_use_reader(sc_context_t *ctx, void * pcsc_context_handle, void * pcsc_c
-		sc_log(ctx, "Reusing the reader");
+		sc_log(ctx, "Try Reusing the reader");
+		struct pcsc_private_data *priv = NULL;
 		sc_reader_t *reader = list_get_at(&ctx->readers, 0);
```
fixed

dengert commented:
@jozsefd https://github.com/OpenSC/OpenSC/pull/2518/commits/da5952e19f4139742f138b2ec6242ec9f3f1a0fb should fix the last problem. It includes an updated version of the diff I sent in comment previously. 

Reader name changed from "REINER SCT cyberJack RFID komfort HUN USB 1" to "OMNIKEY AG Smart Card Reader USB 0"
The "REINER SCT cyberJack RFID komfort HUN" appears to have a pin pad reader but the "OMNIKEY AG Smart Card Reader"
does not and may have other differences too thus the need for this PR.

jozsefd commented:
> OK that debug log shows the first reader was not cleaned up so second call use_reader tried to reuse it again. 
Yes, it seems that there is a reentrant call for pcsc_use_reader().

> But was it with or without the diff I sent?
No, I fetched the diff from the PR.

> Will look at it when my internet comes back up.
Thank you, I hope we can the proper solution for this issue.

jozsefd commented:
> @jozsefd [da5952e](https://github.com/OpenSC/OpenSC/commit/da5952e19f4139742f138b2ec6242ec9f3f1a0fb) should fix the last problem. It includes an updated version of the diff I sent in comment previously.
> 
> Reader name changed from "REINER SCT cyberJack RFID komfort HUN USB 1" to "OMNIKEY AG Smart Card Reader USB 0" The "REINER SCT cyberJack RFID komfort HUN" appears to have a pin pad reader but the "OMNIKEY AG Smart Card Reader" does not and may have other differences too thus the need for this PR.

Thank you @dengert this solution works perfectly when I switch a card between readers:
```
P:9848; T:16012 2022-03-09 14:07:24.933 [cardmod] CardReadFile
P:9848; T:16012 2022-03-09 14:07:24.934 [cardmod] pszDirectoryName = mscp, pszFileName = cmapfile, dwFlags = 0, pcbData=00F5E0E0, ppbData=00F5E0E4
P:9848; T:16012 2022-03-09 14:07:24.934 [cardmod] check_reader_status for CardReadFile
P:9848; T:16012 2022-03-09 14:07:24.934 [cardmod] MD_Function:check_card_status:396 called
P:9848; T:16012 2022-03-09 14:07:24.935 [cardmod] MD_Function:check_card_status:406 returning with: 0x00000000
P:9848; T:16012 2022-03-09 14:07:24.935 [cardmod] sizeof(size_t):4 sizeof(ULONG_PTR):4 sizeof(__int3264):4 sizeof pCardData->hSCardCtx:4
P:9848; T:16012 2022-03-09 14:07:24.935 [cardmod] pCardData->hSCardCtx:0xCD030002 hScard:0xEA010001
P:9848; T:16012 2022-03-09 14:07:24.935 [cardmod] HANDLES CHANGED from  vs->hSCardCtx:0xCD010003 vs->hScard:0xEA020000
P:9848; T:16012 2022-03-09 14:07:24.936 [cardmod] ctx.c:895:sc_ctx_use_reader: called
P:9848; T:16012 2022-03-09 14:07:24.936 [cardmod] reader-pcsc.c:2529:pcsc_use_reader: called
P:9848; T:16012 2022-03-09 14:07:24.936 [cardmod] Try Reusing the reader
P:9848; T:16012 2022-03-09 14:07:24.936 [cardmod] Reader name changed from "REINER SCT cyberJack RFID komfort HUN USB 1" to "OMNIKEY AG Smart Card Reader USB 0"
P:9848; T:16012 2022-03-09 14:07:24.937 [cardmod] reader-pcsc.c:2612:pcsc_use_reader: returning with: -1100 (Generic reader error)
P:9848; T:16012 2022-03-09 14:07:24.937 [cardmod] sc_ctx_use_reader returned -1100
P:9848; T:16012 2022-03-09 14:07:24.937 [cardmod] disassociate_card
P:9848; T:16012 2022-03-09 14:07:24.937 [cardmod] sc_pkcs15_unbind
P:9848; T:16012 2022-03-09 14:07:24.938 [cardmod] pkcs15.c:1317:sc_pkcs15_unbind: called
P:9848; T:16012 2022-03-09 14:07:24.938 [cardmod] pkcs15-pin.c:863:sc_pkcs15_pincache_clear: called
P:9848; T:16012 2022-03-09 14:07:24.938 [cardmod] sc_disconnect_card
P:9848; T:16012 2022-03-09 14:07:24.938 [cardmod] card.c:414:sc_disconnect_card: called
P:9848; T:16012 2022-03-09 14:07:24.939 [cardmod] card-nqApplet.c:186:nqapplet_finish: called
P:9848; T:16012 2022-03-09 14:07:24.939 [cardmod] card-nqApplet.c:191:nqapplet_finish: returning with: 0 (Success)
P:9848; T:16012 2022-03-09 14:07:24.939 [cardmod] card.c:436:sc_disconnect_card: returning with: 0 (Success)
P:9848; T:16012 2022-03-09 14:07:24.939 [cardmod] MD_Function:associate_card:7151 called
P:9848; T:16012 2022-03-09 14:07:24.940 [cardmod] ctx.c:895:sc_ctx_use_reader: called
P:9848; T:16012 2022-03-09 14:07:24.940 [cardmod] reader-pcsc.c:2529:pcsc_use_reader: called
P:9848; T:16012 2022-03-09 14:07:24.940 [cardmod] Freeing reader
P:9848; T:16012 2022-03-09 14:07:24.940 [cardmod] Probing PC/SC reader
P:9848; T:16012 2022-03-09 14:07:24.941 [cardmod] Adding new PC/SC reader 'OMNIKEY AG Smart Card Reader USB 0'
P:9848; T:16012 2022-03-09 14:07:24.941 [cardmod] OMNIKEY AG Smart Card Reader USB 0 check
P:9848; T:16012 2022-03-09 14:07:24.941 [cardmod] current  state: 0x000F0122
P:9848; T:16012 2022-03-09 14:07:24.941 [cardmod] previous state: 0x00000000
P:9848; T:16012 2022-03-09 14:07:24.942 [cardmod] card present, changed
P:9848; T:16012 2022-03-09 14:07:24.943 [cardmod] reader-pcsc.c:1155:detect_reader_features: called
P:9848; T:16012 2022-03-09 14:07:24.943 [cardmod] Requesting reader features ... 
P:9848; T:16012 2022-03-09 14:07:24.944 [cardmod] reader-pcsc.c:2612:pcsc_use_reader: returning with: 0 (Success)
P:9848; T:16012 2022-03-09 14:07:24.945 [cardmod] sc_ctx_get_reader_count(ctx): 1
P:9848; T:16012 2022-03-09 14:07:24.945 [cardmod] card.c:254:sc_connect_card: called
P:9848; T:16012 2022-03-09 14:07:24.945 [cardmod] reader-pcsc.c:609:pcsc_connect: called
P:9848; T:16012 2022-03-09 14:07:24.945 [cardmod] OMNIKEY AG Smart Card Reader USB 0 check
P:9848; T:16012 2022-03-09 14:07:24.946 [cardmod] reader-pcsc.c:387:refresh_attributes: returning with: 0 (Success)
P:9848; T:16012 2022-03-09 14:07:24.947 [cardmod] matching configured ATRs
P:9848; T:16012 2022-03-09 14:07:24.947 [cardmod] matching built-in ATRs
... [initialization log omitted] ...
P:9848; T:16012 2022-03-09 14:07:26.864 [cardmod] MD virtual file system initialized; OPENSSL_VERSION_NUMBER 0x101010CF
P:9848; T:16012 2022-03-09 14:07:26.864 [cardmod] MD_Function:reinit_card:353 returning with: 0x00000000
P:9848; T:16012 2022-03-09 14:07:26.865 [cardmod] MD_Function:reinit_card_for:389 returning with: 0x00000000
P:9848; T:16012 2022-03-09 14:07:26.865 [cardmod] sc.c:340:sc_detect_card_presence: called
P:9848; T:16012 2022-03-09 14:07:26.865 [cardmod] reader-pcsc.c:470:pcsc_detect_card_presence: called
P:9848; T:16012 2022-03-09 14:07:26.866 [cardmod] OMNIKEY AG Smart Card Reader USB 0 check
P:9848; T:16012 2022-03-09 14:07:26.867 [cardmod] reader-pcsc.c:387:refresh_attributes: returning with: 0 (Success)
P:9848; T:16012 2022-03-09 14:07:26.867 [cardmod] reader-pcsc.c:478:pcsc_detect_card_presence: returning with: 5
P:9848; T:16012 2022-03-09 14:07:26.867 [cardmod] sc.c:351:sc_detect_card_presence: returning with: 5
P:9848; T:16012 2022-03-09 14:07:26.867 [cardmod] check_reader_status r=5 flags 0x00000005
P:9848; T:16012 2022-03-09 14:07:26.868 [cardmod] MD_Function:check_card_reader_status:481 returning with: 0x00000000
P:9848; T:16012 2022-03-09 14:07:26.868 [cardmod] MD virtual file system: found 'mscp' directory
P:9848; T:16012 2022-03-09 14:07:26.868 [cardmod] MD virtual file system: found 'cmapfile' file
P:9848; T:16012 2022-03-09 14:07:26.869 [cardmod] returns 'cmapfile' content:
P:9848; T:16012 2022-03-09 14:07:26.869 [cardmod] --- 00FB71A0:1032
P:9848; T:16012 2022-03-09 14:07:26.869 [cardmod]  0000  34003700 63003000 65006600 38006100  2D003400 39003500 35002D00 61003300
P:9848; T:16012 2022-03-09 14:07:26.869 [cardmod]  0020  62003300 2D003300 36003200 30002D00  35003600 31006600 61003300 65003200
P:9848; T:16012 2022-03-09 14:07:26.870 [cardmod]  0040  31003300 62006600 00000000 00000000  03000000 000C3700 64003800 66006100
P:9848; T:16012 2022-03-09 14:07:26.870 [cardmod]  0060  32003500 66002D00 36003500 30006100  2D003900 66006500 32002D00 35003800
P:9848; T:16012 2022-03-09 14:07:26.870 [cardmod]  0080  32006400 2D006200 37006500 34003600  37003700 32003400 66003100 36000000
P:9848; T:16012 2022-03-09 14:07:26.870 [cardmod]  00A0  00000000 00000100 0000000C 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.871 [cardmod]  00C0  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.871 [cardmod]  00E0  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.871 [cardmod]  0100  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.871 [cardmod]  0120  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.872 [cardmod]  0140  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.872 [cardmod]  0160  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.872 [cardmod]  0180  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.873 [cardmod]  01A0  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.873 [cardmod]  01C0  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.873 [cardmod]  01E0  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.873 [cardmod]  0200  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.874 [cardmod]  0220  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.874 [cardmod]  0240  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.874 [cardmod]  0260  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.875 [cardmod]  0280  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.875 [cardmod]  02A0  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.875 [cardmod]  02C0  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.875 [cardmod]  02E0  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.876 [cardmod]  0300  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.876 [cardmod]  0320  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.876 [cardmod]  0340  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.876 [cardmod]  0360  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.877 [cardmod]  0380  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.877 [cardmod]  03A0  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.877 [cardmod]  03C0  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.878 [cardmod]  03E0  00000000 00000000 00000000 00000000  00000000 00000000 00000000 00000000
P:9848; T:16012 2022-03-09 14:07:26.878 [cardmod]  0400  00000000 00000000 
P:9848; T:16012 2022-03-09 14:07:26.878 [cardmod] MD_Function:CardReadFile:4218 returning with: 0x00000000

```

I am about to repeat this test with Citrix VDI, it could take a little longer.

dengert commented:
@jozsefd 
Can you run with two identical readers? 

The  pcsc_new_reader probes the reader using:  `SCardGetAttrib(... SCARD_ATTR_DEVICE_SYSTEM_NAME_A, ...)` to get the name.
For example  the first or only OMNIKEY reader would return "OMNIKEY AG Smart Card Reader USB 0"  but a second would return "OMNIKEY AG Smart Card Reader USB 1" for the other.

For the recoded, what type cards are you using?
Are any of the contactless cards AKA NFC and RFID? 
The "REINER SCT cyberJack RFID komfort HUN USB 1" looks like it can do both contact and contactless.

jozsefd commented:
> @jozsefd I forced pushed a rebased version of this PR. Can you try it some time.

I have double checked it: it works as expected with multiple card readers!
(I am currently having technical difficulties to try it in a Citrix environment)

dengert commented:
Good to hear. I think I have reproduced your environment, but with two different readers, 'SCM Microsystems Inc. SCR35xx v2.0 USB SC Reader 0' and ''Feitian SCR301 0'.  But I see pcsc_use_reader being called and handles changing, but for the same card context thew underlying reader does not change. I see when I remove a card from the reader, Windows does what I think it should do and call "CardDeleteContext".

There could be at least 2 differences.
- Your reader is not reporting that a card has been removed from one or both of your readers. Where mine does. 

- differences in Windows versions. 
What version of Windows are you using? 
I am using: Edition	Windows 10 Pro
Version	21H1
Installed on	‎8/‎3/‎2020
OS build	19043.1586
Experience	Windows Feature Experience Pack 120.2212.4170.0
Windows Update says I am up to date. 

- Any virtual machines involved?
- Different cards, but that should not make a difference.
 
I registered a PIV card ATR to use OpenSC minidriver.  I also removed from the registry entry:  "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\Calais\PIV Device ATR Cache"  the ATR for my PIV card so it would not use the windows PIV driver. I also cleaned up the cert store for and certificates that where from the PIV card in the past when used by tej builtin PIV driver.  

With MD log debugging and OpenSC debug=3 I can insert a PIV card  but never see the code in this PR used because the reader never changes for a pCardData, but handled do, as multiple processes some with multiple tasks have open session to the card.

jozsefd commented:
> Good to hear. I think I have reproduced your environment, but with two different readers, 'SCM Microsystems Inc. SCR35xx v2.0 USB SC Reader 0' and ''Feitian SCR301 0'. But I see pcsc_use_reader being called and handles changing, but for the same card context thew underlying reader does not change. I see when I remove a card from the reader, Windows does what I think it should do and call "CardDeleteContext".
> 

Yes, as I mentioned it is not trivial to reproduce this case, you need an application which keeps running and opens card containers with the CryptoAPI. Even in such application, the case "HANDLES CHANGED" would not always cause a problem, just after some time. If the application is terminated, the CardDeleteContext will be called anyway. (e.g.: certuril -scinfo)

First noticed this problem during Windows Unlocking with Smartcard, but later managed to reproduce it with a certificate loader. 
Please see a sample application at https://github.com/jozsefd/CertLoader (SLN file for VS19 included)

This application tries to emulate what happens during logon. It polls the card readers and if it finds a new card then loads the certificates from it (and prints the subject/issuer data). The application uses the Win32 API CryptAcquireContext to open the card container and CryptReleaseContext to finish accessing it (and no CardDeleteContext in my system).

The application reports the following error without this PR:

`Error accessing card container: Failed to get default context! - 2148073494`

(The error code is: NTE_BAD_KEYSET)

> There could be at least 2 differences.
> 
>     * Your reader is not reporting that a card has been removed from one or both of your readers. Where mine does.
> 

The removal of the card is reported.

>     * differences in Windows versions.
>       What version of Windows are you using?
>       I am using: Edition	Windows 10 Pro
>       Version	21H1
>       Installed on	‎8/‎3/‎2020
>       OS build	19043.1586
>       Experience	Windows Feature Experience Pack 120.2212.4170.0
>       Windows Update says I am up to date.
> 

The testsystem is a German Windows 10 Pro, 21H1 (Build 19043.1586)
Windows Feature Experience Pack 120.2212.4170.0


>     * Any virtual machines involved?
> 

No, this is a bare-metal test system, but I have seen this issue in many other emvironment, including VMs.

>     * Different cards, but that should not make a difference.
> 

This could be really a difference, I am checking if I can send you some test cards.

> 
> I registered a PIV card ATR to use OpenSC minidriver. I also removed from the registry entry: "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\Calais\PIV Device ATR Cache" the ATR for my PIV card so it would not use the windows PIV driver. I also cleaned up the cert store for and certificates that where from the PIV card in the past when used by tej builtin PIV driver.
> 
> With MD log debugging and OpenSC debug=3 I can insert a PIV card but never see the code in this PR used because the reader never changes for a pCardData, but handled do, as multiple processes some with multiple tasks have open session to the card.

As far as I can understand, BaseCSP Caching could be the reason of the issue with "HANDLES CHANGED". The BaseCSP cache happens to be at "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\Calais\Cache". If I clean this cache after removing the card then the pcsc_use_reader would work event without this PR.

I made another experiment with this caching. I implemented a variant which produced a reader specific CardID in the minidriver:
https://github.com/jozsefd/OpenSC/blob/d792e8a5c8f050eca4623a14043e48e09770dd3a/src/minidriver/minidriver.c#L5352-L5432

This reader specific CardID (constructed from the device name and from the real card-id) could also prevent the issue during "Reusing the reader", so it seems like BaseCSP Caching uses solely CardID as the key, even after the device has been changed.

dengert commented:
You have done a lot more research on this problem then I have done!

I would say this PR is ready to be merged. It fixes a corner case that most users will never see and maintains the functionality of [#111246f1d23f6705368d1e615b2a7b3b930a189a](https://github.com/OpenSC/OpenSC/commit/111246f1d23f6705368d1e615b2a7b3b930a189a) when the reader is not changed.


The OpenSC minidriver was written to support 7.07 and is usable by baseCSP and KSP. 

I also see https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptacquirecontexta says CryptAcquireContextA is deprecated. I see your test: https://github.com/jozsefd/CertLoader/blob/master/CertLoader/CertLoader.cpp is using CryptAcquireContext. 

I don't know (and are not directly applicable for OpenSC)  but maybe in your environment:
- Is CryptAcquireContext deprecated or just the "A" version. 
- what should be used in its place. 
- does login use CSP or KSP
- what other windows components only support baseCSP.
- does the use of Cryptography\Calais\Cache change if used by KSP

I stumbled on this article that might be related to the Citrix problem. A bit old but Citrix has been around a long time. It adds a time component to how long things are cached. I would assume newer versions of Windows have this.

https://support.microsoft.com/en-us/topic/a-smart-card-logon-to-a-terminal-session-stops-responding-server-that-is-running-windows-server-2008-and-windows-server-2008-r2-cbbd51ce-86b9-c537-4884-50a0a1ca7826 

This too:
https://docs.microsoft.com/en-us/windows/security/identity-protection/smart-cards/smart-card-debugging-information#debugging-and-tracing-using-wpp

jozsefd commented:
> You have done a lot more research on this problem then I have done!
> 

I have recognized this issue during implementing a new card driver (s. card_nqApplet.c) and I kept trying to find the reason for it.

> I would say this PR is ready to be merged. It fixes a corner case that most users will never see and maintains the functionality of [#111246f1d23f6705368d1e615b2a7b3b930a189a](https://github.com/OpenSC/OpenSC/commit/111246f1d23f6705368d1e615b2a7b3b930a189a) when the reader is not changed.
> 

I agree, nice job!

> The OpenSC minidriver was written to support 7.07 and is usable by baseCSP and KSP.
> 
> I also see https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptacquirecontexta says CryptAcquireContextA is deprecated. I see your test: https://github.com/jozsefd/CertLoader/blob/master/CertLoader/CertLoader.cpp is using CryptAcquireContext.
> 

I have added NCrypt (KSP) support to my test tool. It also fails with the minidriver w/o this PR, but with a different error:
Error accessing card container: NCryptEnumKeys failed - 0x8010000d 

In the minidriver logs it happens at the same place: it fails with "Card not present" after "Reusing the reader".

> I don't know (and are not directly applicable for OpenSC) but maybe in your environment:
> 
>     * Is CryptAcquireContext deprecated or just the "A" version.
> 

It is the legacy Crypto API, both "A" and "W" are deprecated for new software, but it is still supported by Windows.

>     * what should be used in its place.
> 
>     * does login use CSP or KSP
> 
>     * what other windows components only support baseCSP.
> 
>     * does the use of Cryptography\Calais\Cache change if used by KSP
> 

I suppose the unterlaying functionality is the same both API, KSP adds a more versatile interface and new features like EC support.

> 
> I stumbled on this article that might be related to the Citrix problem. A bit old but Citrix has been around a long time. It adds a time component to how long things are cached. I would assume newer versions of Windows have this.
> 
> https://support.microsoft.com/en-us/topic/a-smart-card-logon-to-a-terminal-session-stops-responding-server-that-is-running-windows-server-2008-and-windows-server-2008-r2-cbbd51ce-86b9-c537-4884-50a0a1ca7826
> 
> This too: https://docs.microsoft.com/en-us/windows/security/identity-protection/smart-cards/smart-card-debugging-information#debugging-and-tracing-using-wpp

My Citrix account still not available, hope to get it fixed soon so I can do some tests.

frankmorgner commented:
Great to hear that you made it work, good job! What I find confusing is the context that is carried back and forth between minidriver and reader driver.

If I understand correctly, you want to do the following:
detect whether we have new handles
- if the new handles are belonging to a new reader:
1. initialize the reader object with the new handles
2. clean the cache in the minidriver
- if the new handles are belonging to the old reader:
1. initialize the reader object with the new handles
2. don't touch the cache in the minidriver

I think a simpler approach would leave the whole context in the calling minidriver by
1. adding a new callback `pcsc_check_reder_handles(sc_context_t *ctx, sc_reader_t *reader, void * pcsc_context_handle, void * pcsc_card_handle)`, where [most of your changes of this PR go](https://github.com/OpenSC/OpenSC/pull/2518/files#diff-5ec6a4c0bb7d2a6d20faeb402457d27e79f482cbab5fb92587660c4ed5317c7aR2556-R2565).
2. in minidriver, call `pcsc_check_reader_handles` and depending on its return value (i.e. whether we have handles for a new reader or not), either call `pcsc_use_reader()` as is today (i.e. when it's the same reader), or call `reinit_card_for()` (i.e. when this are handles for a different reader).

If I'm not mistaken, this allows keeping all the context handling in minidriver and it avoids using `SC_CTX_FLAG_TERMINATE` and all existing callbacks in reader-pcsc.c can stay as they are today, just one new function is added. Am I missing something?

dengert commented:
reinit_card_for  calls disassociate that may try and do some card operation such as logout or lock using the the old handles that should not exist or the new handles. This is very similar to a forked unix process cleaning up. The SC_CTX_FLAG_TERMINATE was added for PKCS11.  

later during associate, pcsc_use_reader is called to use the new handles At that point the SC_CTX_FLAG_TERMINATE indicates that the old reader needs to be deleted too before adding the new reader. 

If you would like to rewrite this, feel free to do so. 

@frankmorgner  In addition to these changes to minidriver, see  https://github.com/arekinath/PivApplet/issues/64 where they are trying to use PIV-like token but with EC keys. This is turning out to be a nightmare, and there maybe issues with minidriver with non-PIV cards but with ECDSA keys and login.  I am spending too much time on minidriver.

frankmorgner commented:
> reinit_card_for calls disassociate that may try and do some card operation such as logout or lock using the the old handles that should not exist or the new handles. This is very similar to a forked unix process cleaning up. The SC_CTX_FLAG_TERMINATE was added for PKCS11.
> 
> later during associate, pcsc_use_reader is called to use the new handles At that point the SC_CTX_FLAG_TERMINATE indicates that the old reader needs to be deleted too before adding the new reader.

This is complicated. First, I'd like to avoid back and forth messeging through OpenSC's layers by using different flags, because we as developers will eventually loose track of all this complexity.

Second, SC_CTX_FLAG_TERMINATE is related to `sc_context_t`, it is not (only) reader specific. Indeed this flag would be appropriate if we would choose to completely re-initialize all internal handles of the minidriver if the given SCARDCONTEXT/SCARDHANDLE are changing. However, we are calling sc_context_create/sc_release_context only during CardAcquireContext/CardDeleteContext.

> If you would like to rewrite this, feel free to do so.

done here https://github.com/OpenSC/OpenSC/pull/2525

dengert commented:
@jozsefd  Can you try @frankmorgner 's https://github.com/OpenSC/OpenSC/pull/2525 to see if it fixes your problem?

jozsefd commented:
> @jozsefd Can you try @frankmorgner 's #2525 to see if it fixes your problem?

For sure, I am about to test it.

jozsefd commented:
> @jozsefd Can you try @frankmorgner 's #2525 to see if it fixes your problem?

#2525 works properly, see my test results at https://github.com/OpenSC/OpenSC/pull/2525#issuecomment-1069390152

frankmorgner commented:
I merged https://github.com/OpenSC/OpenSC/pull/2525 instead as it has fewer changes and uses a simpler logic for messaging between the abstraction layers
